hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaTargetManagement.java

```diff
@@ -271,38 +280,48 @@ public void deleteTargets(final Long... targetIDs) {
+    public Long countTargetByFilters(final Collection<TargetUpdateStatus> status, final Boolean overdueState,
+                                     final String searchText, final Long installedOrAssignedDistributionSetId,
+                                     final Boolean selectTargetWithNoTag, final String... tagNames) {
+        final List<Specification<JpaTarget>> specList = buildSpecificationList(
+                new FilterParams(installedOrAssignedDistributionSetId, status, overdueState, searchText,
+                        selectTargetWithNoTag, tagNames),
+                true);
         return countByCriteriaAPI(specList);
     }
 
-    private static List<Specification<JpaTarget>> buildSpecificationList(final Collection<TargetUpdateStatus> status,
-            final String searchText, final Long installedOrAssignedDistributionSetId,
-            final Boolean selectTargetWithNoTag, final boolean fetch, final String... tagNames) {
-        final List<Specification<JpaTarget>> specList = new LinkedList<>();
-        if (status != null && !status.isEmpty()) {
-            specList.add(TargetSpecifications.hasTargetUpdateStatus(status, fetch));
+    private static List<Specification<JpaTarget>> buildSpecificationList(final FilterParams filterParams,
+                                                                         final boolean fetch) {
+        final List<Specification<JpaTarget>> specList = new ArrayList<>();
+        if (filterParams.getFilterByStatus() != null && !filterParams.getFilterByStatus().isEmpty()) {
+            specList.add(TargetSpecifications.hasTargetUpdateStatus(filterParams.getFilterByStatus(), fetch));
+        }
+        if (filterParams.getOverdueState() != null) {
+            specList.add(
+                    TargetSpecifications.isOverdue(new TimestampCalculator().calculateOverdueTimestamp()));
         }
-        if (installedOrAssignedDistributionSetId != null) {
+        if (filterParams.getFilterByDistributionId() != null) {
             specList.add(
-                    TargetSpecifications.hasInstalledOrAssignedDistributionSet(installedOrAssignedDistributionSetId));
+                    TargetSpecifications
+                            .hasInstalledOrAssignedDistributionSet(filterParams.getFilterByDistributionId()));
         }
-        if (!Strings.isNullOrEmpty(searchText)) {
-            specList.add(TargetSpecifications.likeNameOrDescriptionOrIp(searchText));
+        if (!Strings.isNullOrEmpty(filterParams.getFilterBySearchText())) {
+            specList.add(TargetSpecifications.likeNameOrDescriptionOrIp(filterParams.getFilterBySearchText()));
         }
-        if (selectTargetWithNoTag != null && (selectTargetWithNoTag || (tagNames != null && tagNames.length > 0))) {
-            specList.add(TargetSpecifications.hasTags(tagNames, selectTargetWithNoTag));
+        if (filterParams.getSelectTargetWithNoTag() != null && (filterParams.getSelectTargetWithNoTag()
```

Comment:
This expression is complex, better extract to private method and provide a proper name what this expression is about